url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/SSLSocketChannel2.java

```diff
@@ -0,0 +1,380 @@
+		bufferallocations++;
+	}
+
+	public int write( ByteBuffer src ) throws IOException {
+		if( !isHandShakeComplete() ) {
+			processHandshake();
+			return 0;
+		}
+		// assert ( bufferallocations > 1 ); //see #190
+		if( bufferallocations <= 1 ) {
+			createBuffers( sslEngine.getSession() );
+		}
+		int num = socketChannel.write( wrap( src ) );
+		return num;
+
+	}
+
+	/**
+	 * Blocks when in blocking mode until at least one byte has been decoded.<br>
+	 * When not in blocking mode 0 may be returned.
+	 * 
+	 * @return the number of bytes read.
+	 **/
+	public int read( ByteBuffer dst ) throws IOException {
+		if( !dst.hasRemaining() )
+			return 0;
+		if( !isHandShakeComplete() ) {
+			if( isBlocking() ) {
+				while ( !isHandShakeComplete() ) {
+					processHandshake();
+				}
+			} else {
+				processHandshake();
+				if( !isHandShakeComplete() ) {
+					return 0;
+				}
+			}
+		}
+		// assert ( bufferallocations > 1 ); //see #190
+		if( bufferallocations <= 1 ) {
+			createBuffers( sslEngine.getSession() );
+		}
+		/* 1. When "dst" is smaller than "inData" readRemaining will fill "dst" with data decoded in a previous read call.
+		 * 2. When "inCrypt" contains more data than "inData" has remaining space, unwrap has to be called on more time(readRemaining)
+		 */
+		int purged = readRemaining( dst );
+		if( purged != 0 )
+			return purged;
+
+		/* We only continue when we really need more data from the network.
+		 * Thats the case if inData is empty or inCrypt holds to less data than necessary for decryption
+		 */
+		assert ( inData.position() == 0 );
+		inData.clear();
+
+		if( !inCrypt.hasRemaining() )
+			inCrypt.clear();
+		else
+			inCrypt.compact();
+
+		if( isBlocking() || readEngineResult.getStatus() == Status.BUFFER_UNDERFLOW )
+			if( socketChannel.read( inCrypt ) == -1 ) {
+				return -1;
+			}
+		inCrypt.flip();
+		unwrap();
+
+		int transfered = transfereTo( inData, dst );
+		if( transfered == 0 && isBlocking() ) {
+			return read( dst ); // "transfered" may be 0 when not enough bytes were received or during rehandshaking
+		}
+		return transfered;
+	}
+	/**
+	 * {@link #read(ByteBuffer)} may not be to leave all buffers(inData, inCrypt)
+	 **/
+	private int readRemaining( ByteBuffer dst ) throws SSLException {
+		if( inData.hasRemaining() ) {
+			return transfereTo( inData, dst );
+		}
+		if( !inData.hasRemaining() )
+			inData.clear();
+		// test if some bytes left from last read (e.g. BUFFER_UNDERFLOW)
+		if( inCrypt.hasRemaining() ) {
+			unwrap();
+			int amount = transfereTo( inData, dst );
+			if( amount > 0 )
+				return amount;
+		}
+		return 0;
+	}
+
+	public boolean isConnected() {
+		return socketChannel.isConnected();
+	}
+
+	public void close() throws IOException {
+		sslEngine.closeOutbound();
+		sslEngine.getSession().invalidate();
+		if( socketChannel.isOpen() )
+			socketChannel.write( wrap( emptybuffer ) );// FIXME what if not all bytes can be written
+		socketChannel.close();
+		exec.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)